README
########################################################################
#            Software Engineering Build Server Information
#
#  Written by:  Michael Jennings
#               Ted T'so
#
#  $Id: README,v 1.4 2000/06/14 00:28:17 mej Exp $
#
########################################################################

0.  INTRODUCTION AND QUICK START
--------------------------------

This document is intended as a brief overview of the new Software
Engineering build system:  how it works, how to use it, and some of the
design philosophies behind it.  It also details the various standards
policies which govern its contents, so those planning on adding to
and/or maintaining part(s) of the tree should read those sections
carefully.

Before you can build any products, you must set your machine up to be
able to run the build system tools.  Since you are reading this README
file, you must have checked out the "base" module.  Within that module
is a directory called "pm" which contains another directory called
"getopt."  This getopt directory contains the Getopt::Mixed Perl
module which is required for the build scripts to function properly.
You will need to go into pm/getopt and run the following command as
root:  "perl Makefile.PL ; make ; make install".  Once you have done
this, return to the base directory and proceed as shown below.

For those wishing to simply build a particular VA product, you will
find a directory called products under this one.  It contains a series
of files with .prod extensions.  Find the file with the name of the
product you wish to build, then invoke the following command at your
shell prompt:

$ ./buildtool -p <product>

...where <product> is the name of the product file, but without the
.prod on the end.  This command should be run by your regular user
account (not root) and should be run from the directory which contains
buildtool and this README file.  Executing this command will build all
packages associated with that product, both binary and source packages.
If you wish to build ISO images as well, use -bi as well as -p.

The following sections will have more detailed information on how to
use the system, so if you are going to be involved in the actual day-
to-day maintenance of the system or any packages within the system,
keep reading!


1.  OVERALL PHILOSOPHY
----------------------

The build system is designed on two basic concepts:  the "product" and
the "package."  A "product" is a collection of packages which form some
coherent entity when you put them all together, such as a distribution.
A "package" is the basic unit which the build system must know how to
take from its "source" form and generate its "target" form.  Take care
not to confuse this term with "package file."  Although a package can
take the form of a package file, like an SRPM, not all packages are
package files, nor are all package files packages.

In general, the build system works with two general classifications of
packages.  Some packages are modified by VA before being included in
our enhanced distribution(s), while others are taken directly from
RedHat or some other source.  The build system must be able to
seamlessly handle both these cases.  For the latter case, the packages
are simply kept as SRPM's, and buildtool knows how to explode the SRPM
and build from the exploded sources.

The former case is much more complicated.  Packages which are modified
internally by VA are stored as directory trees containing the source,
the necessary spec files, and any other files required by the build
process.  These directory trees will hereafter be referred to by their
proper term in revision control lingo, "modules."

Several tools are employed as user-level interfaces to this system.
The first, called "revtool," handles all interaction with the revision
control system.  It encapsulates all revision-control-system-specific
commands and gives them an interface which is independent of the
underlying system.  Thus, if the repository was moved from CVS to,
say, BitKeeper, revtool's code would change, but its interface would
remain the same.

The second tool is "buildtool," mentioned briefly above.  buildtool is
what takes a product name/version pair, like "va-redhat-6.2.1" and
translates that via a series of product definition/configuration files
("prod files" for short) into a set of packages to build.  It then
builds those packages, creating both source and binary package files.
It can also use these packages to build ISO images for product CD's.

There are other tools, such as pkgtool and safetool, which are not yet
ready for widespread use.  We'll discuss those when that situation
changes.


2.  REVISION CONTROL WITH REVTOOL
---------------------------------

The syntax of revtool is intended to be generic, in that it should
apply well to any arbitrary revision control system one might choose.
Each function revtool performs should translate into zero or more
commands specific to the revision control system in question.  For
example, a "put" in revtool-speak is analogous to a "commit" in CVS
lingo.  A "get" in revtool is either a "checkout" or an "update" in
CVS, or a "pull" in BitKeeper.  Thus, revtool acts as an encapsulation
layer for the underlying revision control system.

revtool also assists in routine upkeep actions, like maintaining the
ChangeLog.  revtool will append your commit log entry to the ChangeLog
automatically.  It will even create the ChangeLog for you if it
doesn't exist, and the date and identity information are provided for
you.

Finally, revtool acts as a means of policy enforcement.  This allows
us to ensure that commits are properly tagged, for example.
Consistency and cleanliness of operation are vital for the ongoing
functionality, stability, and manageability of the build system, and
revtool allows us to enforce the policies that provide those things
before anything enters the tree that could disrupt it.



3.  BUILDING PRODUCTS
---------------------

4.  PACKAGE MAINTAINER'S GUIDE
------------------------------

5.  STANDARDS AND CONVENTIONS
-----------------------------


ChangeLog
Tue Sep 25 09:40:00 2001                        Michael Jennings (mej)

Fixes for the erasing of the RPM's just built by mzbuild. :P
----------------------------------------------------------------------
Fri Oct  5 16:11:25 2001                        Michael Jennings (mej)

Assorted fixes.
----------------------------------------------------------------------
Fri Oct  5 17:34:53 2001                        Michael Jennings (mej)

Get rid of that buffy stuff.
----------------------------------------------------------------------
Wed Oct 10 15:07:17 2001                        Michael Jennings (mej)

Fixes to make imgtool work properly.
----------------------------------------------------------------------
Sun Nov  4 12:40:34 2001                        Michael Jennings (mej)

Fixes for mediatool and friends.
----------------------------------------------------------------------
Sun Nov  4 20:47:26 2001                        Michael Jennings (mej)

Let's try this again.
----------------------------------------------------------------------
Sun Nov  4 20:48:32 2001                        Michael Jennings (mej)

Scripts work better when they are executable....
----------------------------------------------------------------------
Mon Nov  5 12:02:22 2001                        Michael Jennings (mej)

New tool for making bootable CD's.
----------------------------------------------------------------------
Thu Nov 29 13:52:06 2001                        Michael Jennings (mej)

Lots of changes over the past month.... :-)
----------------------------------------------------------------------
Thu Dec 20 19:25:40 2001                        Michael Jennings (mej)

The beginnings of some SGML documentation.
----------------------------------------------------------------------
Wed Jan  2 19:12:58 2002                        Michael Jennings (mej)

Debian packaging goop.  Yikes.
----------------------------------------------------------------------
Thu Jan  3 10:12:16 2002                        Michael Jennings (mej)

Cleaned up the Debian stuff thanks to some help from Brian Almieda
<bma@debian.org>.
----------------------------------------------------------------------
Thu Jan  3 10:16:44 2002                        Michael Jennings (mej)

Make sure to include debian/ directory in source tarball.
----------------------------------------------------------------------
Fri Jan  4 18:36:49 2002                        Michael Jennings (mej)

Working on deb support.
----------------------------------------------------------------------
Mon Feb 18 15:38:36 2002                        Michael Jennings (mej)

Added support for passing SRPM's to mzmerge.
----------------------------------------------------------------------
Mon Feb 25 18:49:59 2002                        Michael Jennings (mej)

Argh.  How long has that extra 's' on the end of query_log() been
there? :P
----------------------------------------------------------------------
Mon Feb 25 19:00:20 2002                        Michael Jennings (mej)

Updated Debian changelog
----------------------------------------------------------------------
Tue Feb 26 15:49:08 2002                        Michael Jennings (mej)

Fixed the usage of $EDITOR in case it's more than one word.
----------------------------------------------------------------------
Tue Feb 26 15:58:10 2002                        Michael Jennings (mej)

If arguments are given to mzput without -a or -r, assume -x, but do
the ChangeLog.
----------------------------------------------------------------------
Tue Feb 26 16:47:34 2002                        Michael Jennings (mej)

Better finding of ChangeLog...I hope.
----------------------------------------------------------------------
Tue Feb 26 16:51:05 2002                        Michael Jennings (mej)

Cosmetic changes.  I'm testing the changelog finder thingie. :)
----------------------------------------------------------------------
Tue Feb 26 16:52:17 2002                        Michael Jennings (mej)

Still testing...
----------------------------------------------------------------------
Tue Feb 26 16:58:39 2002                        Michael Jennings (mej)

More testing.
----------------------------------------------------------------------
Tue Feb 26 17:01:21 2002                        Michael Jennings (mej)

Argh...it broke.
----------------------------------------------------------------------
Tue Feb 26 17:03:03 2002                        Michael Jennings (mej)

Still fixing stuff...
----------------------------------------------------------------------
Tue Feb 26 17:05:14 2002                        Michael Jennings (mej)

Okay, let's see if *this* works too....
----------------------------------------------------------------------
Wed Feb 27 16:06:26 2002                        Michael Jennings (mej)

Fixed a bug with commits from the top level directory when specifying
filenames.
----------------------------------------------------------------------
Sat May  4 09:36:06 2002                        Michael Jennings (mej)

Syncing back up....
----------------------------------------------------------------------
Mon Jun 24 13:30:47 2002                        Michael Jennings (mej)

Fixes for the new rpm.
----------------------------------------------------------------------
Wed Jul 10 00:04:41 2002                        Michael Jennings (mej)

Fix handling of binary-only packages.
----------------------------------------------------------------------
Wed Jul 10 00:11:14 2002                        Michael Jennings (mej)

Fix hierarchical variable fallbacks.
----------------------------------------------------------------------
Wed Jul 10 00:22:09 2002                        Michael Jennings (mej)

Taking a crack at the apt stuff.
----------------------------------------------------------------------
Wed Jul 10 11:44:52 2002                        Michael Jennings (mej)

apt stuff should work now.

mod/Deb.pm
# Mezzanine DEB Perl Module
# 
# Copyright (C) 2001-2007, Michael Jennings
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies of the Software, its documentation and marketing & publicity
# materials, and acknowledgment shall be given in the documentation, materials
# and software packages that this Software was used.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
# $Id: Deb.pm,v 1.8 2011/06/25 19:41:34 mej Exp $
#

package Mezzanine::Deb;

BEGIN {
    use strict;
    use Exporter   ();
    use Mezzanine::Util;
    use vars ('$VERSION', '@ISA', '@EXPORT', '@EXPORT_OK', '%EXPORT_TAGS');

    # set the version for version checking
    $VERSION     = 2.1;

    @ISA         = ('Exporter');
    # Exported functions go here
    @EXPORT      = ();
    %EXPORT_TAGS = ( );

    # Exported variables go here
    @EXPORT_OK   = ('$VERSION');
}
use vars ('@EXPORT_OK');

### Private global variables

### Initialize exported package variables

# Constants

### Initialize private global variables

### Function prototypes

# Private functions

### Module cleanup
END {
}

### Function definitions

sub
deb_form_command
{
    my $type = shift;
    my $cmd;

    $type = "" if (!defined($type));

    if (&pkgvar_topdir()) {
        $cmd = "cd " . &pkgvar_topdir() . " && ";
    } else {
        $cmd = "";
    }
    if ($type eq "build") {
        if (! &pkgvar_command()) {
            &pkgvar_command("dpkg-buildpackage");
        }
        $cmd .= &pkgvar_command();
        if (&pkgvar_buildroot()) {
            $cmd .= " --buildroot=\"" . &pkgvar_buildroot() . "\"";
        }
    } elsif ($type eq "install") {
        if (! &pkgvar_command()) {
            &pkgvar_command("dpkg");
        }
        $cmd .= &pkgvar_command();
        if (&pkgvar_instroot()) {
            $cmd .= " --root=\"" . &pkgvar_instroot() . "\"";
        }
    } elsif ($type eq "contents") {
    } elsif ($type eq "files") {
    } elsif ($type eq "query") {
    }
    if (&pkgvar_parameters()) {
        $cmd .= " " . &pkgvar_parameters();
    }
    dprint "Command:  $cmd\n";
    return $cmd;
}

sub
deb_install
{
    my $cmd;
    local *DPKG;

    if (! &pkgvar_filename()) {
        return (MEZZANINE_SYNTAX_ERROR, "No package specified for install");
    }
    if (&pkgvar_subtype() eq "sdeb") {
        $cmd = &deb_form_command("install") . " -x " . &pkgvar_filename();
    } else {
        $cmd = &deb_form_command("install") . " -i " . &pkgvar_filename();
    }
    if (!open(DPKG, "$cmd 2>&1 |")) {
        eprint "Execution of \"$cmd\" failed -- $!\n";
    }
    while (<DPKG>) {
        print;
    }
    close(DPKG);
    if ($? != 0) {
        return (MEZZANINE_UNSPECIFIED_ERROR, &pkgvar_command() . " returned " . ($? >> 8));
    }
    return (MEZZANINE_SUCCESS, &pkgvar_filename() . " successfully installed.");
}

sub
deb_show_contents
{
    my ($cmd, $line);
    my @results;
    local *DPKG;

    if (! &pkgvar_filename()) {
        return (MEZZANINE_SYNTAX_ERROR, "No package specified for query");
    }
    $cmd = &deb_form_command("contents") . " -c " . &pkgvar_filename();
    if (!open(DPKG, "$cmd 2>&1 |")) {
        eprint "Execution of \"$cmd\" failed -- $!\n";
    }
    @results = <DPKG>;
    close(DPKG);
    return ($? >> 8, @results);
}

sub
deb_list_files
{
    my ($cmd, $line);
    my @results;
    local *DPKG;

    if (! &pkgvar_filename()) {
        return (MEZZANINE_SYNTAX_ERROR, "No package specified for query");
    }
    $cmd = &deb_form_command("files") . " -c " . &pkgvar_filename();
    if (!open(DPKG, "$cmd 2>&1 |")) {
        eprint "Execution of \"$cmd\" failed -- $!\n";
    }
    @results = <DPKG>;
    close(DPKG);
    return ($? >> 8, @results);
}

sub
deb_query
{
    my $query_type = $_[0];
    my ($cmd, $line);
    my (@prov, @deps);
    local *DPKG;

    $cmd = &deb_form_command("query") . " -I";
    if ($query_type eq "d") {
    $cmd = "$dpkg -I " . ($pkg_file ? "$pkg_file" : "");
    dprint "About to run \"$cmd\"\n";
    if (!open(DPKG, "$cmd 2>&1 |")) {
        eprint "Execution of \"$cmd\" failed -- $!\n";
    }
    while (<DPKG>) {
        chomp($line = $_);
        if ($query_type eq "d") {
            if ($line =~ /^\s*Provides:\s+(.*)$/) {
                ($line = $1) =~ s/[\(\)]//g;
                push @prov, split(/\s*,\s*/, $line);
            } elsif ($line =~ /^\s*Depends:\s+(.*)$/) {
                ($line = $1) =~ s/[\(\)]//g;
                push @deps, split(/\s*,\s*/, $line);
            }
        }
    }
    close(DPKG);
    dprint "\"$cmd\" returned $?\n";
    if ($? != 0) {
        return MEZZANINE_UNSPECIFIED_ERROR;
    }
    foreach my $prov (@prov) {
        print "Capability:  $prov\n";
    }
    foreach my $dep (@deps) {
        print "Dependency:  $dep\n";
    }   
    } else {
        eprint "Unrecognized query type \"$query_type\"\n";
        return MEZZANINE_SYNTAX_ERROR;
    }
    return MEZZANINE_SUCCESS;
}

sub
deb_build
{

}


### Private functions

1;
